File: packages/awesome_notifications_localizations/android/src/main/kotlin/me/carda/awesome_notifications_localizations/LocalizationManager.kt
package me.carda.awesome_notifications_localizations

import android.content.Context

/**
 * Stores the current language code (persisted in SharedPreferences). Mirrors the
 * original AndroidAwnCore LocalizationManager: default is the system locale,
 * normalized to lowercase.
 */
class LocalizationManager private constructor() {
    companion object {
        private const val PREFS = "awn_localizations"
        private const val KEY = "languageCode"

        @Volatile
        private var instance: LocalizationManager? = null

        fun getInstance(): LocalizationManager =
            instance ?: synchronized(this) {
                instance ?: LocalizationManager().also { instance = it }
            }
    }

    private var appContext: Context? = null

    fun init(context: Context) {
        appContext = context.applicationContext
    }

    fun setLocalization(languageCode: String?): Boolean {
        val context = appContext ?: return false
        val code = (languageCode ?: systemLanguage(context)).lowercase()
        prefs(context).edit().putString(KEY, code).apply()
        return true
    }

    fun getLocalization(): String {
        val context = appContext ?: return "en"
        val fallback = systemLanguage(context)
        return prefs(context).getString(KEY, fallback) ?: fallback
    }

    private fun systemLanguage(context: Context): String {
        @Suppress("DEPRECATION")
        val locale = context.resources.configuration.locale
        return locale.toLanguageTag().lowercase()
    }

    private fun prefs(context: Context) =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
}

File: packages/awesome_notifications_localizations/ios/awesome_notifications_localizations/Sources/awesome_notifications_localizations/LocalizationManager.swift
import Foundation

/// Stores the current language code (persisted in UserDefaults). Mirrors the
/// original IosAwnCore LocalizationManager: default is the system language,
/// normalized to lowercase.
final class LocalizationManager {
    static let shared = LocalizationManager()
    private init() {}

    private let storageKey = "awn_localizations_languageCode"

    @discardableResult
    func setLocalization(_ languageCode: String?) -> Bool {
        let code = (languageCode ?? systemLanguage()).lowercased()
        UserDefaults.standard.set(code, forKey: storageKey)
        return true
    }

    func getLocalization() -> String {
        return UserDefaults.standard.string(forKey: storageKey) ?? systemLanguage()
    }

    private func systemLanguage() -> String {
        return (Locale.preferredLanguages.first ?? "en").lowercased()
    }
}
